src/net_processing.cpp

```diff
@@ -75,6 +75,8 @@ static constexpr int64_t INBOUND_PEER_TX_DELAY = 2 * 1000000;
 static constexpr int64_t GETDATA_TX_INTERVAL = 60 * 1000000;
 /** Maximum delay (in microseconds) for transaction requests to avoid biasing some peers over others. */
 static constexpr int64_t MAX_GETDATA_RANDOM_DELAY = 2 * 1000000;
+/** How long to wait (in microseconds) before expiring a getdata request to a peer */
+static constexpr int64_t TX_EXPIRY_INTERVAL = 10 * GETDATA_TX_INTERVAL;
```

Comment:
Does the use of 10 here imply that-- assuming all peers INV a transaction to us at roughly the same time-- that we get no more robustness to transaction suppression than if we only had ten peers?  If so, perhaps this should be equal to the maximum number of peers times the interval

Reply:
No, `TX_EXPIRY_INTERVAL` only removes tx's that are already in the `in_flight` list, which means we've already sent a GETDATA for them at least `TX_EXPIRY_INTERVAL` microseconds ago.

Reply:
However, I think there's another bug that causes problems here -- if you have two offers with nearby `m_process_time` values, then you'll hit the `else` clause of `last_request_Time <= nNow - GETDATA_TX_INTERVAL` and call `RequestTx(txid)` which will just return because it's already present in `m_tx_announced`, and then you'll clear `inv.hash` (ie `txid`) from `tx_process_time`, and never actually request it from that peer. So I think this also needs:

                } else {
                    // This transaction is in flight from someone else; queue
                    // up processing to happen after the download times out
                    // (with a slight delay for inbound peers, to prefer
                    // requests to outbound peers).
    +               state.m_tx_download.m_tx_announced.erase(inv.hash);
                    RequestTx(&state, txid, nNow);
                }

Reply:
Oh, that only works for the request side. If we get:

     0s peer A: INV deadbeef...
     1s peer B: INV deadbeef...
     ...
     5s peer Z: INV deadbeef

and then we happen to query some unresponsive peers, we'll see:

    10s: -> peer D: GETDATA deadbeef...
    610s: [expired]
    611s: -> peer Q: GETDATA deadbeef...
    1211s: [expired]

At which point every peer will have decided "it's been >15m, I'll expire the deadbeef... tx from mapRelay, and then every node will respond with NOTFOUND for that tx...

Reply:
> Does the use of 10 here imply that-- assuming all peers INV a transaction to us at roughly the same time-- that we get no more robustness to transaction suppression than if we only had ten peers?

@gmaxwell I'm not sure I understand the question -- is the concern that an adversary could keep re-announcing a transaction to us, and every 10 minutes we'll retry that adversary instead of going to our other peers?

I think that does seem problematic, but since we prefer outbound peers over inbound ones, we should at least be able to query all our outbound peers before an adversary would be able to cause us to re-cycle back to them.  I could, as you suggest, raise this to 125 minutes to eliminate this problem, but that seemed like an excessive time to me that we might (potentially) stop transaction relay.  Of course in theory this code shouldn't really kick in except for misbehaving peers, so I don't have a strong intuition of where to set it.... Thoughts?

Reply:
@ajtowns Thanks for catching that bug around `RequestTx()`.  I'll be including a fix the next time I update this PR.

> Oh, that only works for the request side. If we get:
> 
> ```
>  0s peer A: INV deadbeef...
>  1s peer B: INV deadbeef...
>  ...
>  5s peer Z: INV deadbeef
> ```
> and then we happen to query some unresponsive peers, we'll see:
> 
> ```
> 10s: -> peer D: GETDATA deadbeef...
> 610s: [expired]
> 611s: -> peer Q: GETDATA deadbeef...
> 1211s: [expired]
> ```
> At which point every peer will have decided "it's been >15m, I'll expire the deadbeef... tx from mapRelay, and then every node will respond with NOTFOUND for that tx...

I think there's a misunderstanding here.  The design here is supposed to be that every 1 minute (GETDATA_TX_INTERVAL), we will send a GETDATA for a txid that we want to a new peer, while every 10 minutes (TX_EXPIRY_INTERVAL) we'll clear out the in-flight request to a peer we've asked for a given txid, to make room for asking that peer for other txid's (since we cap the in-flight size to 100).  Your example makes me think that you were looking at the 10-minute timer as applying to both new requests and expiring from the data structure, which should not be the case...?

Reply:
Yeah, I think I got the `GetTxRequestTime` check confused with checking whether it was inflight via another node at all, rather than just hadn't been asked for for a while.

Reply:
My comment on the interval was mostly that we'll expire all past the first ten to offer to us before we ever get to them. The discussion with AJ sorted out my confusion.